File: pkgs/split/split.c
/*
 * Mini Object Storage, (C) 2014 Minio, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <sys/param.h>
#include <sys/types.h>

#include <ctype.h>
#include <fcntl.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include <assert.h>
#include "split.h"

static
size_t _get_filesize(int fd)
{
        struct stat st;
        assert(fstat(fd, &st) != -1);
        return (ssize_t) st.st_size;
}

static
ssize_t read_write_chunk(int in, int out, ssize_t bytes)
{
        ssize_t n, m;
        char *buf = NULL;

        buf = calloc(bytes, 1);
        assert(buf != NULL);

        n = read(in, buf, bytes);
        if (n < 0)
                return -1;

        m = write(out, buf, n);
        if (m < 0)
                return -1;

        if (buf)
                free(buf);
        return m;
}

static int
_allocate_newchunk(char *chunkname)
{

        int chunk = -1;

        if (!chunkname)
                return -1;

        if ((strlen(chunkname)) >= MAXPATHLEN) {
                fprintf (stderr, "chunkname + suffix too long");
                return -1;
        }

        chunk = open (chunkname, O_CREAT|O_WRONLY, S_IRUSR|S_IWUSR);
        return chunk;
}

/*
 * Generate chunks by chunking at input bytes.
 */
int
minio_split(char *filename, ssize_t bytecnt)
{
	ssize_t fsize;
        int output = -1;
        int input = -1;
        int remainder = 0;
        char newchunk[MAXPATHLEN] = {0,};
        int chunk_size = 0;
        int i = 0;

        if (bytecnt < 0)
                return -1;

        if ((input = open(filename, O_RDONLY)) < 0)
                return -1;

        fsize = _get_filesize(input);
        remainder = fsize % bytecnt;

        if (remainder == 0)
                chunk_size = fsize / bytecnt;
        else
                chunk_size = (fsize + (bytecnt - remainder)) / bytecnt;

        if (chunk_size == 0)
                return -1;

        for (i = 0; i < chunk_size; i++) {
                snprintf (newchunk, sizeof(newchunk)-1, "%s.%d", filename, i);

                if ((output = _allocate_newchunk(newchunk)) < 0)
                        return -1;

                if (read_write_chunk(input, output, bytecnt) < 0)
                        return -1;
        }
        return 0;
}

File: pkg/donut/split/split.go
/*
 * Minimalist Object Storage, (C) 2014 Minio, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package split

import (
	"bufio"
	"bytes"
	"errors"
	"io"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

// Message - message structure for results from the Stream goroutine
type Message struct {
	Data []byte
	Err  error
}

// Stream reads from io.Reader, splits the data into chunks, and sends
// each chunk to the channel. This method runs until an EOF or error occurs. If
// an error occurs, the method sends the error over the channel and returns.
// Before returning, the channel is always closed.
//
// The user should run this as a gorountine and retrieve the data over the
// channel.
//
//  channel := make(chan Message)
//  go Stream(reader, chunkSize, channel)
//  for chunk := range channel {
//    log.Println(chunk.Data)
//  }
func Stream(reader io.Reader, chunkSize uint64) <-chan Message {
	ch := make(chan Message)
	go splitStreamGoRoutine(reader, chunkSize, ch)
	return ch
}

func splitStreamGoRoutine(reader io.Reader, chunkSize uint64, ch chan Message) {
	// we read until EOF or another error
	var readError error

	// run this until an EOF or error occurs
	for readError == nil {
		// keep track of how much data has been read
		var totalRead uint64
		// Create a buffer to write the current chunk into
		var bytesBuffer bytes.Buffer
		bytesWriter := bufio.NewWriter(&bytesBuffer)
		// read a full chunk
		for totalRead < chunkSize && readError == nil {
			var currentRead int
			// if we didn't read a full chunk, we should attempt to read again.
			// We create a byte array representing how much space is left
			// unwritten in the given chunk
			chunk := make([]byte, chunkSize-totalRead)
			currentRead, readError = reader.Read(chunk)
			// keep track of how much we have read in total
			totalRead = totalRead + uint64(currentRead)
			// prune the array to only what has been read, write to chunk buffer
			chunk = chunk[0:currentRead]
			bytesWriter.Write(chunk)
		}
		// flush stream to underlying byte buffer
		bytesWriter.Flush()
		// if we have data available, send it over the channel
		if bytesBuffer.Len() != 0 {
			ch <- Message{bytesBuffer.Bytes(), nil}
		}
	}
	// if we have an error other than an EOF, send it over the channel
	if readError != io.EOF {
		ch <- Message{nil, readError}
	}
	// close the channel, signaling the channel reader that the stream is complete
	close(ch)
}

// JoinFiles reads from a given directory, joins data in chunks with prefix and sends
// an io.Reader.
//
//  var err error
//  for err == nil {
//     buf := make([]byte, 1024*1024)
//     reader := JoinFiles("mydirectory", "mypreferred-prefix")
//     _, err = reader.Read(buf)
//     fmt.Println(buf)
// }
//
func JoinFiles(dirname string, inputPrefix string) (io.Reader, error) {
	reader, writer := io.Pipe()
	fileInfos, readError := ioutil.ReadDir(dirname)
	if readError != nil {
		writer.CloseWithError(readError)
		return nil, readError
	}

	var newfileInfos []os.FileInfo
	for _, fi := range fileInfos {
		if strings.Contains(fi.Name(), inputPrefix) == true {
			newfileInfos = append(newfileInfos, fi)
		}
	}

	if len(newfileInfos) == 0 {
		nofilesError := errors.New("no files found for given prefix " + inputPrefix)
		writer.CloseWithError(nofilesError)
		return nil, nofilesError
	}

	go joinFilesGoRoutine(newfileInfos, writer)
	return reader, nil
}

func joinFilesGoRoutine(fileInfos []os.FileInfo, writer *io.PipeWriter) {
	for _, fileInfo := range fileInfos {
		file, err := os.Open(fileInfo.Name())
		defer file.Close()
		for err != nil {
			writer.CloseWithError(err)
			return
		}
		_, err = io.Copy(writer, file)
		if err != nil {
			writer.CloseWithError(err)
			return
		}
	}
	writer.Close()
}

// FileWithPrefix - Takes a file and splits it into chunks with size chunkSize. The output
// filename is given with outputPrefix.
func FileWithPrefix(filename string, chunkSize uint64, outputPrefix string) error {
	// open file
	file, err := os.Open(filename)
	defer file.Close()
	if err != nil {
		return err
	}

	if outputPrefix == "" {
		return errors.New("Invalid argument outputPrefix cannot be empty string")
	}

	// start stream splitting goroutine
	ch := Stream(file, chunkSize)

	// used to write each chunk out as a separate file. {{outputPrefix}}.{{i}}
	i := 0

	// write each chunk out to a separate file
	for chunk := range ch {
		if chunk.Err != nil {
			return chunk.Err
		}
		err := ioutil.WriteFile(outputPrefix+"."+strconv.Itoa(i), chunk.Data, 0600)
		if err != nil {
			return err
		}
		i = i + 1
	}
	return nil
}
